aspnetcore-webapi/RateLimitingDemo/RateLimitingDemo.UsingCustomMiddleware/Middlewares/RateLimitingMiddleware.cs

```diff
@@ -0,0 +1,96 @@
+﻿using Microsoft.Extensions.Caching.Distributed;
+using RateLimitingDemo.UsingCustomMiddleware.Decorators;
+using RateLimitingDemo.UsingCustomMiddleware.Extensions;
+using System.Net;
+
+namespace RateLimitingDemo.UsingCustomMiddleware.Middlewares;
+
+public class RateLimitingMiddleware
+{
+    private readonly RequestDelegate _next;
+
+    private readonly IDistributedCache _cache;
+    public RateLimitingMiddleware(RequestDelegate next, IDistributedCache cache)
+    {
+        _next = next;
+        _cache = cache;
+    }
+
+    public async Task InvokeAsync(HttpContext context)
+    {
+        var endpoint = context.GetEndpoint();
+
+
+        var rateLimitingDecorator = endpoint?.Metadata.GetMetadata<LimitRequests>();
+
+
+        if (rateLimitingDecorator is null)
+        {
+            await _next(context);
+            return;
+        }
+
+        var key = GenerateClientKey(context);
+
+        var clientStatistics = await GetClientStatisticsByKey(key);
+
+
+        if (clientStatistics != null)
+        {
+
+            if (DateTime.UtcNow < clientStatistics.LastSuccessfulResponseTime.AddSeconds(rateLimitingDecorator.TimeWindow) && clientStatistics.NumberOfRequestsCompletedSuccessfully == rateLimitingDecorator.MaxRequests)
+            {
+                context.Response.StatusCode = (int)HttpStatusCode.TooManyRequests;
+                return;
+            }
+        }
+
+        await UpdateClientStatisticsStorage(key, rateLimitingDecorator.MaxRequests);
+
+        await _next(context);
```

Comment:
Some formatting issues here, mainly newlines, but it might be whitespace as well since it has problems in the article snippet as well.

Reply:
fixed formatting issue. Will do the same in article snippet as well.